Spigot-Server-Patches/0594-Significantly-improve-performance-of-the-end-generat.patch

```diff
@@ -0,0 +1,74 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SuperCoder7979 <25208576+SuperCoder7979@users.noreply.github.com>
+Date: Tue, 3 Nov 2020 23:48:05 -0600
+Subject: [PATCH] Significantly improve performance of the end generation
+
+This patch implements a noise cache for the end which significantly reduces the computation time of generation. This results in about a 3x improvement.
+
+Original code by SuperCoder7979 and Gegy in Lithium, licensed under LGPL-3.0 (Source: https://github.com/jellysquid3/lithium-fabric)
+
+Co-authored-by: Gegy <gegy1000@gmail.com>
+Co-authored-by: Dylan Xaldin <Puremin0rez515@gmail.com>
+Co-authored-by: pop4959 <pop4959@gmail.com>
+
+diff --git a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+index 9c15cd303ccc2928be4e3e7080e74e8314d7e0f2..3604f815ab3df190e0582f279d3164af8b6f6880 100644
+--- a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
++++ b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+@@ -3,6 +3,9 @@ package net.minecraft.server;
+ import com.google.common.collect.ImmutableList;
+ import com.mojang.serialization.Codec;
+ import com.mojang.serialization.codecs.RecordCodecBuilder;
++import it.unimi.dsi.fastutil.HashCommon; // Paper
++import it.unimi.dsi.fastutil.objects.Reference2ObjectMap; // Paper
++import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap; // Paper
+ import java.util.List;
+ 
+ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+@@ -22,6 +25,16 @@ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+     private final BiomeBase k;
+     private final BiomeBase l;
+     private final BiomeBase m;
++    // Paper start
++    private static final class NoiseCache {
++        public long[] keys = new long[8192];
++        public float[] values = new float[8192];
++        public NoiseCache() {
++            java.util.Arrays.fill(keys, Long.MIN_VALUE);
++        }
++    }
++    private static final ThreadLocal<Reference2ObjectMap<NoiseGenerator3Handler, NoiseCache>> noiseCache = ThreadLocal.withInitial(Reference2ObjectOpenHashMap::new);
```

Comment:
Not really sure that a ThreadLocal is the best idea, especially if you have worlds loading/unloading, any reason why this is preferred over just using a standard map/array?

There's no real garauntees over how threads will be handled, e.g. if we get parallel, will it be using multiple threads which are killed afterwards, etc

Could just store it in a singular array per manager, that way you can "guarantee" enough safety there

Reply:
If we have an instance of WorldChunkManagerTheEnd per world, that should unload memory as a world unloads.

Though thread safety SHOULD be provided by the world gen system under CURRENT design as a single world is only being generated by a single thread at a time.

The only case I can see this mattering is if this noise cache is shared across multiple worlds.

Reply:
Our code used ThreadLocal as the backing lossy cache isn't thread safe. Even though the generation isn't done concurrently, we wanted to be extra sure we wouldn't mess anything up. While it's not the best solution, it gets the job done 😅

Reply:
It doesn't hurt to be future proof. I was more responding to Cats concern over world loading.

As long as my understanding is correct that we have an instance per world and its not shared between worlds, there should be no concerns over memory leak and I think this current impl is good to go.

Reply:
Yea, I didn't realize that the locals where stored on the Thread, so should be cleaned up when the thread is dead, so that's a non-issue, at this point, lgtm

Reply:
I do think `ThreadLocal` is the best choice here to be as absolutely safe as possible while remaining performant.

I was able to confirm after doing heap analysis that a memory leak does exist when unloading and loading an end dimension as the worker threads never die. I was able to easily resolve this issue by swapping the `Reference2ObjectMap` with a `WeakHashMap` so that the GC will cleanup no longer in use noise generator keys. Thankfully, it remains just as performant as the original PR description results.